fem/gridfunc.cpp

```diff
@@ -431,6 +431,186 @@ void GridFunction::GetVectorValue(int i, const IntegrationPoint &ip,
    }
 }
 
+double GridFunction::GetBdrValue(int i, const IntegrationPoint &ip, int vdim)
+const
+{
+   Array<int> dofs;
+   fes->GetBdrElementDofs(i, dofs);
+   fes->DofsToVDofs(vdim-1, dofs);
+   Vector DofVal(dofs.Size()), LocVec;
+   const FiniteElement *be = fes->GetBE(i);
+   MFEM_ASSERT(be, "boundary FE does not exist");
+   MFEM_ASSERT(be->GetMapType() == FiniteElement::VALUE,
+               "invalid boundary FE map type");
+   be->CalcShape(ip, DofVal);
+   GetSubVector(dofs, LocVec);
+
+   return (DofVal * LocVec);
+}
+
+void GridFunction::GetBdrVectorValue(int i, const IntegrationPoint &ip,
+                                     Vector &val) const
+{
+   const FiniteElement *be = fes->GetBE(i);
+   int dof = be->GetDof();
+   Array<int> vdofs;
+   fes->GetBdrElementVDofs(i, vdofs);
+   Vector loc_data;
+   GetSubVector(vdofs, loc_data);
+   if (be->GetRangeType() == FiniteElement::SCALAR)
+   {
+      MFEM_ASSERT(be->GetMapType() == FiniteElement::VALUE,
+                  "invalid boundary FE map type");
+      Vector shape(dof);
+      be->CalcShape(ip, shape);
+      int vdim = fes->GetVDim();
+      val.SetSize(vdim);
+      for (int k = 0; k < vdim; k++)
+      {
+         val(k) = shape * ((const double *)loc_data + dof * k);
+      }
+   }
+   else
+   {
+      int spaceDim = fes->GetMesh()->SpaceDimension();
+      DenseMatrix vshape(dof, spaceDim);
+      ElementTransformation *Tr = fes->GetBdrElementTransformation(i);
+      Tr->SetIntPoint(&ip);
+      be->CalcVShape(*Tr, vshape);
+      val.SetSize(spaceDim);
+      vshape.MultTranspose(loc_data, val);
+   }
+}
+
+double GridFunction::GetFaceValue(int i, int side, const IntegrationPoint &ip,
+                                  int vdim, int *pdir) const
+{
```

Comment:
There seem to be a lot of code duplication in `GetFaceValue`, `GetFaceVectorValue`,  and the two `GetFaceValues`. Is there a way to consolidate it?

Reply:
Yes, it is just about the choice of the face, but I introduced an auxiliary private function for that in 3b7ff80 replacing this procedure even in the existing methods.